### PR TITLE
CMake quality of life changes

### PR DESCRIPTION
**Pull request by samuelpmish**
see https://github.com/EnzymeAD/Enzyme/issues/981

This PR aims to improve CMake support by defining/exporting some targets that allow users to integrate Enzyme into their project by only adding two lines of code to an existing CMakeLists.txt.

The two idiomatic ways to add a dependency (like Enzyme) to a CMake project are:

1. `add_subdirectory` (usable with git submodules, CMake's `FetchContent`)
```cmake
add_subdirectory(/path/to/Enzyme)
...
target_link_libraries(my_enzyme_target PUBLIC Enzyme)
```

2. `find_package` (assumes Enzyme was built ahead of time and `make install`ed or user specified `Enzyme_DIR` to CMake)

```cmake
find_package(Enzyme REQUIRED)
...
target_link_libraries(my_enzyme_target PUBLIC Enzyme)
```

-------------

At present, Enzyme does not support this kind of CMake integration, which means users end up manually managing compilation and link flags, see [for example](https://github.com/EnzymeAD/CMake-Template/blob/f0ea23f928a39af453b41773efe6a6d84e96315f/CMakeLists.txt#L16-L23).

Instead, this PR adds some interface targets that move that flag management out of user projects and into Enzyme. 

I'm currently testing against the following simple CMake project (with `add_subdirectory` and `find_package`):
[enzyme_consumer_project.zip](https://github.com/EnzymeAD/Enzyme/files/11134092/enzyme_consumer_project.zip)
and LLVM versions 14 and 15.

--------------

I still need some clarification about how Enzyme is intended to be used in a CMake project: I have seen two ways:

1. `ClangEnzyme`:
```
/path/to/clang++ -flegacy-pass-manager -Xclang -load -Xclang /path/to/ClangEnzyme-14.so
```

2. `LLDEnzyme`
```
/path/to/clang++ -fuse-ld=lld -Wl,--lto-legacy-pass-manager -Wl,-mllvm=-load=/path/to/LLDEnzyme-14.so ...
```

For now, I define separate targets for each of these approaches, and users can pick whichever approach they prefer by doing
either
```cmake
target_link_libraries(my_enzyme_target PUBLIC ClangEnzyme)
# or
target_link_libraries(my_enzyme_target PUBLIC LLDEnzyme)
```

### CODE REVIEW

**Comment by samuelpmish** [comment]
Also, it seems the flags required for each target differ slightly by LLVM version, so I need some help understanding what to do for versions older that LLVM14.

And when I build with LLVM15, the LLDEnzyme approach, the CMake integration appears to be working correctly (it's inserting the flags I expect it to), but I'm getting an undefined symbol error:

```
/path/to/llvm/15/build/bin/clang++ -fuse-ld=lld -Wl,-mllvm=-load=/path/to/LLDEnzyme-15.so main.cpp.o -o lldenzyme_test
[build] ld.lld: error: undefined symbol: __enzyme_autodiff(void*, double)
```

File: enzyme/Enzyme/CMakeLists.txt [hunk 1]
```diff
@@ -283,6 +283,29 @@ set_target_properties(LLDEnzyme-${LLVM_VERSION_MAJOR} PROPERTIES
         LINK_FLAGS "-Wl,-flat_namespace -Wl,-undefined -Wl,suppress")
 endif()
 
+add_library(LLDEnzyme INTERFACE)
+target_compile_options(LLDEnzyme INTERFACE -flto)
+target_link_options(LLDEnzyme INTERFACE "SHELL: -fuse-ld=lld")
+if(${LLVM_VERSION_MAJOR} LESS 15)
```
Comment (wsmoses) [inline enzyme/Enzyme/CMakeLists.txt:289]: This is only if version > 12 and < 15

File: enzyme/Enzyme/CMakeLists.txt [hunk 2]
```diff
@@ -283,15 +283,20 @@ set_target_properties(LLDEnzyme-${LLVM_VERSION_MAJOR} PROPERTIES
...
+
 add_library(LLDEnzyme INTERFACE)
 target_compile_options(LLDEnzyme INTERFACE -flto)
 target_link_options(LLDEnzyme INTERFACE "SHELL: -fuse-ld=lld")
-if(${LLVM_VERSION_MAJOR} LESS 15) 
+
+set(LTO_LEGACY_MANAGER_VERSIONS 13 14)
+if (${LLVM_VERSION_MAJOR} IN_LIST LTO_LEGACY_MANAGER_VERSIONS)
   target_link_options(LLDEnzyme INTERFACE "SHELL: -Wl,--lto-legacy-pass-manager")
 endif()
```
Comment (wsmoses) [inline enzyme/Enzyme/CMakeLists.txt:294]: Per your comment on the PR itself -- I think the LTO plugin also needs to use a different (new passmanger specific) flag on 15+. I don't know what it is offhand, but presumably we can find this in the LLVM docs.

**Comment by samuelpmish** [comment]
I'm trying this out on my mac, and seeing some new error messages (even with LLVM 14):

```
/path/to/llvm/14/build/bin/clang++ -O3 -DNDEBUG -arch arm64 -isysroot 
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk 
-Wl,-search_paths_first -Wl,-headerpad_max_install_names -fuse-ld=lld -Wl,--lto-legacy-pass-manager 
-Wl,-mllvm -load=/path/to/Enzyme/build/lib/LLDEnzyme-14.dylib 
CMakeFiles/enzyme_only.dir/tests/enzyme_only.cpp.o -o enzyme_only  lib/libgtest_main.a lib/libgtest.a

Error opening '/path/to/Enzyme/build/lib/LLDEnzyme-14.dylib': 
dlopen(/path/to/Enzyme/build/lib/LLDEnzyme-14.dylib, 0x0009): symbol not found in flat namespace (__ZN4llvm17IntervalPartitionC1Ev)
```

@wsmoses any idea if the mac builds need to be set up differently?

**Comment by wsmoses** [comment]
@samuelpmish can you make a version of that test to run on CI? I want to check what it happens across the whole CI tree.

Ignore the 18.04 issues now, we need to resolve that.

**Comment by samuelpmish** [comment]
some test results on my local machine (ubuntu 22.04):

```
running add_subdirectory tests
llvm 11: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 12: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 13: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 14: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 15: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ❌
llvm 16: configuring cmake ... ✅ | building ClangEnzyme test ... ❌ | building LLDEnzyme test ... ❌

running find_package tests
llvm 11: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 12: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 13: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 14: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 15: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ❌
llvm 16: configuring cmake ... ✅ | building ClangEnzyme test ... ❌ | building LLDEnzyme test ... ❌
```

Note: llvm ## corresponds to the version of LLVM enzyme was built with for those tests.

So, LLDEnzyme is still failing w/ LLVM 15, and LLVM16 isn't supported yet. 

Hopefully, I can get access to a mac this week to run these tests on that platform too.

**Comment by wsmoses** [comment]
How does the LLVM15 LLD test fail?

**Comment by samuelpmish** [comment]
```
/home/sam/code/llvm/15/build/bin/clang++ -O3 -DNDEBUG -fuse-ld=lld 
-Wl,-mllvm -Wl,-load=/home/sam/code/Enzyme/fork/enzyme/install/15/lib/LLDEnzyme-15.so 
CMakeFiles/enzyme_test_2.dir/main.cpp.o -o enzyme_test_2 

ld.lld: error: undefined symbol: __enzyme_autodiff(void*, double)
```
and with "-v"
```
/home/sam/code/llvm/15/build/bin/clang++ -O3 -DNDEBUG -fuse-ld=lld 
-Wl,-mllvm -Wl,-load=/home/sam/code/Enzyme/fork/enzyme/install/15/lib/LLDEnzyme-15.so
CMakeFiles/enzyme_test_2.dir/main.cpp.o -o enzyme_test_2 -v
clang version 15.0.7
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/sam/code/llvm/15/build/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/10
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda-12.0, version 

"/home/sam/code/llvm/15/build/bin/ld.lld" -pie -z relro --hash-style=gnu --eh-frame-hdr 
-m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o enzyme_test_2 /lib/x86_64-linux-gnu/Scrt1.o 
/lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/12/crtbeginS.o 
-L/home/sam/code/llvm/15/build/bin/../lib/x86_64-unknown-linux-gnu 
-L/usr/lib/gcc/x86_64-linux-gnu/12 -L/usr/lib/gcc/x86_64-linux-gnu/12/../../../../lib64 
-L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib 
-L/usr/lib -mllvm -load=/home/sam/code/Enzyme/fork/enzyme/install/15/lib/LLDEnzyme-15.so CMakeFiles/enzyme_test_2.dir/main.cpp.o -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc 
/usr/lib/gcc/x86_64-linux-gnu/12/crtendS.o /lib/x86_64-linux-gnu/crtn.o
ld.lld: error: undefined symbol: __enzyme_autodiff(void*, double)
>>> referenced by ld-temp.o
>>>               lto.tmp:(main)
>>> referenced by ld-temp.o
>>>               lto.tmp:(main)
>>> referenced by ld-temp.o
>>>               lto.tmp:(main)
>>> referenced 1 more times
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by samuelpmish** [comment]
does https://github.com/EnzymeAD/Enzyme/pull/1090 relate to getting LLDEnzyme working with LLVM 15?

**Comment by wsmoses** [comment]
Try doing `-load-pass-plugin=` instead of `-load=`

**Comment by samuelpmish** [comment]
same issue with `--load-pass-plugin=` (`lld help` shows it with 2 leading hyphens):

```
"/home/sam/code/llvm/15/build/bin/ld.lld" -pie -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 
-dynamic-linker /lib64/ld-linux-x86-64.so.2 -o enzyme_test_2 /lib/x86_64-linux-gnu/Scrt1.o 
/lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/12/crtbeginS.o 
-L/home/sam/code/llvm/15/build/bin/../lib/x86_64-unknown-linux-gnu 
-L/usr/lib/gcc/x86_64-linux-gnu/12 -L/usr/lib/gcc/x86_64-linux-gnu/12/../../../../lib64 -L/lib/x86_64-linux-gnu 
-L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib -L/usr/lib 
--load-pass-plugin=/home/sam/code/Enzyme/fork/enzyme/install/15/lib/LLDEnzyme-15.so
CMakeFiles/enzyme_test_2.dir/main.cpp.o 
-lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/12/crtendS.o /lib/x86_64-linux-gnu/crtn.o
ld.lld: error: undefined symbol: __enzyme_autodiff(void*, double)
>>> referenced by ld-temp.o
>>>               lto.tmp:(main)
>>> referenced by ld-temp.o
>>>               lto.tmp:(main)
>>> referenced by ld-temp.o
>>>               lto.tmp:(main)
>>> referenced 1 more times
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by samuelpmish** [comment]
here's the corresponding compatibility on an M2 macbook pro:

```
running add_subdirectory tests
llvm 11: configuring cmake ... ✅ | building ClangEnzyme test ... ❌ | building LLDEnzyme test ... ❌
llvm 12: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ❌
llvm 13: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 14: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 15: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ❌
llvm 16: configuring cmake ... ✅ | building ClangEnzyme test ... ❌ | building LLDEnzyme test ... ❌
running find_package tests
llvm 11: configuring cmake ... ✅ | building ClangEnzyme test ... ❌ | building LLDEnzyme test ... ❌
llvm 12: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ❌
llvm 13: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 14: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ✅
llvm 15: configuring cmake ... ✅ | building ClangEnzyme test ... ✅ | building LLDEnzyme test ... ❌
llvm 16: configuring cmake ... ✅ | building ClangEnzyme test ... ❌ | building LLDEnzyme test ... ❌
```

llvm 16 doesn't compile (as expected), but I'd like to resolve the other errors if possible. I'll summarize the build issues below case-by-case:

--------

## LLVM 11

Both `add_subdirectory` and `find_package` versions emit the same errors:
```
In file included from /Users/mish2/code/Enzyme/fork/enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp:14:
In file included from /Users/mish2/code/llvm/11/llvm/include/llvm/ADT/StringExtras.h:16:
In file included from /Users/mish2/code/llvm/11/llvm/include/llvm/ADT/ArrayRef.h:12:
In file included from /Users/mish2/code/llvm/11/llvm/include/llvm/ADT/Hashing.h:47:
In file included from /Users/mish2/code/llvm/11/llvm/include/llvm/Support/DataTypes.h:16:
/Users/mish2/code/llvm/11/llvm/include/llvm-c/DataTypes.h:33:3: error: "The standard header <cstdint> is not C++11 compliant. Must #define __STDC_LIMIT_MACROS before #including llvm-c/DataTypes.h"
/Users/mish2/code/llvm/11/llvm/include/llvm-c/DataTypes.h:38:3: error: "The standard header <cstdint> is not C++11 compliant. Must #define __STDC_CONSTANT_MACROS before #including llvm-c/DataTypes.h"
```
My first attempt was to just conditionally append the compilation flags
```
-D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS
```
to the command line, but the same errors appear when those are included. I'm unsure what to do here.

## LLVM 12

```
[20/20] Linking CXX executable enzyme_test_2                                        
FAILED: enzyme_test_2                                                            
: && /Users/mish2/code/llvm/12/build/bin/clang++ -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -fuse-ld=lld -Wl,-mllvm -Wl,-load=/Users/mish2/code/Enzyme/consumer_projects/test_add_subdirectory/build/enzyme/Enzyme/LLDEnzyme-12.dylib CMakeFiles/enzyme_test_2.dir/main.cpp.o -o enzyme_test_2   && :                   
ld64.lld: warning: ignoring unknown argument: -platform_version                  
ld64.lld: warning: ignoring unknown argument: -search_paths_first                
ld64.lld: warning: ignoring unknown argument: -headerpad_max_install_names       
ld64.lld: warning: -sdk_version is required when emitting min version load command.  Setting sdk version to match provided min version
Cannot open macos: No such file or directory 
```
Similar errors are discussed here: https://lists.llvm.org/pipermail/cfe-dev/2019-March/061666.html
which makes it seem like `lld` support for macs is inconsistent (unclear how to handle this).

## LLVM 13 & 14
No issues!

## LLVM 15

```
/Users/mish2/code/llvm/15/build/bin/clang++ -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -fuse-ld=lld -Wl,-mllvm -Wl,-load=/Users/mish2/code/Enzyme/consumer_projects/test_add_subdirectory/build/enzyme/Enzyme/LLDEnzyme-15.dylib CMakeFiles/enzyme_test_2.dir/main.cpp.o -o enzyme_test_2   && :
ld64.lld: error: undefined symbol: __enzyme_autodiff(void*, double)                 
>>> referenced by /tmp/lto.tmp:(symbol dsquare(double)+0x1c)                        
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

This is consistent with the LLDEnzyme issues on ubuntu, described above [here](https://github.com/EnzymeAD/Enzyme/pull/1084#issuecomment-1506091497).

**Comment by wsmoses** [comment]
@samuelpmish got everything working now. As a bonus, also added an Enzyme_CLANG_EXE which can be used to set and/or verify CMAKE_C_COMPILER is the right one.

As for MacOS, I think what you saw is an LLVM problem, you may be able to resolve this by adding -resource-dir /path/to/resource-dir. You can find this as follows:
```bash
(base) wmoses@31-38-32 ~ % clang -print-resource-dir   
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.0
```

**Pull request merged**